main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -617,6 +617,77 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
             case (b, i, v) => WeededAst.Expression.ArrayStore(b, i, v, mkSL(sp1, sp2))
           }
 
+        case ParsedAst.Expression.NewChannel(sp1, tpe, expOpt, sp2) =>
+          expOpt match {
+            case None =>
+              // Case 1: NewChannel takes no expression that states the buffer size
+              val bufferSize = WeededAst.Expression.Int32(lit = 0, mkSL(sp2, sp2))
+              WeededAst.Expression.NewChannel(bufferSize, Types.weed(tpe), mkSL(sp1, sp2)).toSuccess
+            case Some(exp) =>
+              // Case 2: NewChannel takes an expression that states the buffer size
+              exp match {
+                case l: WeededAst.Expression.Int32 if (l.lit < 0) =>
+                    IllegalBufferSize(mkSL(sp1, sp2)).toFailure
+                case _ =>
+                  visit(exp, unsafe) map {
+                    case e => WeededAst.Expression.NewChannel(e, Types.weed(tpe), mkSL(sp1, sp2))
+                  }
+              }
+          }
+
+        case ParsedAst.Expression.GetChannel(sp1, exp, sp2) =>
+          visit(exp,unsafe) map {
+            case e => WeededAst.Expression.GetChannel(e, mkSL(sp1, sp2))
+          }
+        
+        case ParsedAst.Expression.PutChannel(exp1, exp2, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+          @@(visit(exp1, unsafe), visit(exp2, unsafe)) map {
+            case (e1, e2) => WeededAst.Expression.PutChannel(e1, e2, loc)
+          }
+
+        case ParsedAst.Expression.Spawn(sp1, fn, params, sp2) =>
+          val loc = mkSL(sp1, sp2)
+          val func = WeededAst.Expression.VarOrDef(Name.QName(sp1, Name.RootNS, fn, sp2), loc)
+
+          def createExpression(params: List[ParsedAst.Expression], arguments: List[Name.Ident], paramNumber: Int): Validation[WeededAst.Expression, WeederError] =
```

Comment:
I think this whole block of code can be simplified, but let us talk about it more in detail.

I would validate the code first and only then construct the let-ladder, the lambda, and the apply using mkApplyCurried.

This would avoid the complicate of createExpression which returns a Validation.

Perhaps we can peer program it at some point.